### ISSUE TITLE
Support creation of Complex Networks

### ISSUE THREAD

**Issue by FlorianCassayre** [FlorianCassayre/DeepLearningUtil#3]
Currently the library only allows the creation of basic feedforward networks (essentially a stack of layers stored as a list). But some tasks requires a more complex architecture with mixed connections between layers.
This implementation would need deep internal modifications, namely the class `Network` itself needs to be redesigned and potentially `Layer` aswell.

**Comment by hugo4715** [comment]
Maybe supporting reccurent network could be usefull

**Comment by FlorianCassayre** [comment]
Yes, indeed (although these fit in another very different category).
I was initially thinking about "networks in a network" such as Inception, U-Net and so on.
![Inception](https://indico.io/blog/wp-content/uploads/2016/02/inception_cleaned.png)
![U-Net](https://lmb.informatik.uni-freiburg.de/Publications/2016/CABR16/figureMICCAI2016small.png)

**Comment by FlorianCassayre** [comment]
Basically two layers are needed: `SplitLayer` and `ConcatLayer`
But implementing these will violate the rule where each layer must be independent from the others. Perhaps it's fine and I'm just over-engineering the whole thing.

**Comment by hugo4715** [comment]
I'm not that advanced in this domain, so i don't understand those network topologies (i'm learning all of this by myself and i takes a lot of time) but i think the goal of the library is to support a lot of topologies. If the current framework or rules can't be used to create those specific layers, you may want to change them.
